libraries/AC_Avoidance/AP_Astar_RT.h

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Common/Location.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_HAL/AP_HAL.h>
+#include "AP_OAVisGraph.h"
+#include <AC_Avoidance/AP_OADatabase.h>
+
+#define OA_RTASTAR_NUM_POINTS_PER_OBJECT 8
+#define OA_RTASTAR_DEBUG_ENABLE 0
+
+/*
+ * Real Time A* avoidance algorithm for avoiding the polygon and circular fence and dynamic objects detected by the proximity sensor
+ */
+class AP_OART_AStar {
+public:
+    AP_OART_AStar();
+
+    /* Do not allow copies */
+    AP_OART_AStar(const AP_OART_AStar &other) = delete;
+    AP_OART_AStar &operator=(const AP_OART_AStar&) = delete;
+
+    // send configuration info stored in front end parameters
+    void set_config(float margin_max) { _margin_max_cm = MAX(margin_max, 0.0f)*100.0f;  }
+
+    // update return status enum
+    enum AP_OART_Astar_State : uint8_t {
+        RT_ASTAR_STATE_NOT_REQUIRED = 0,
+        RT_ASTAR_STATE_ERROR,
+        RT_ASTAR_STATE_SUCCESS
+    };
+
+    // run background task to find best path and update avoidance_results
+    // returns true and populates origin_new and destination_new if OA is required.  returns false if OA is not required
+    AP_OART_Astar_State update(const Location& current_loc, const Location& destination, const Vector2f &ground_speed_vec, Location &origin_new, Location &destination_new, bool proximity_only);
+
+    // get singleton instance
+    static AP_OART_AStar *get_singleton() {
+        return _singleton;
+    }
+
+#if OA_RTASTAR_DEBUG_ENABLE
+    void send_debug_info(mavlink_channel_t chan, uint16_t interval_ms);
+#endif
+
+private:
+
+    // various types of errors that can occur
+    enum class AP_OA_RTAStar_Error : uint8_t {
+        RT_ASTAR_ERROR_NONE = 0,             // No error
+        RT_ASTAR_ERROR_OUT_OF_MEMORY,        // Out of memory (mostly happens when arrays can't be expanded)
+        RT_ASTAR_ERROR_DB_ERROR,             // Some error with OA DB
+        RT_ASTAR_ERROR_TOO_MANY_POINTS,      // Too many OA DB obstacles
+        RT_ASTAR_ERROR_NO_POSITION_ESTIMATE, // No position estimate
+        RT_ASTAR_ERROR_TOO_LONG_TIME_TAKEN,  // Can't find path in a reasonable time
+        RT_ASTAR_ERROR_COULD_NOT_FIND_PATH   // Can't find a safe path at all
+    };
+
+    // create nodes from OA DB obstacles. This will attempt to group closesly located obstacles.  Returns false on error or DB is empty
+    bool create_oa_database_nodes();
+
+    // group closesly packed obstacles together. Returns false on error
+    bool group_oa_items(Graph &grouping_graph);
+
+    // do depth first search to figure out connected obstacles. Returns false on error
+    bool do_dfs(uint16_t &iteration, uint16_t start, Vector3f grouped_nodes[OA_RTASTAR_NUM_POINTS_PER_OBJECT], Graph &grouping_graph, bool visited[]);
```

Comment:
maybe rename "do_dfs" to "find_connected_obstacles" or something similar.  The "what" is more important than the "how".